### PR TITLE
push Filter down LeftJoin

### PR DESCRIPTION
**Pull request by czpmango**

### CODE REVIEW

File: src/optimizer/rule/PushFilterDownLeftJoinRule.cpp [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+        return TransformResult::noTransform();
+    }
+
+    TransformResult result;
+    result.eraseAll = true;
+
+    auto* newFilterNode = static_cast<graph::Filter*>(oldFilterNode->clone());
+    newFilterNode->setCondition(objPool->add(filterPicked.release()));
+    newFilterNode->setInputVar(leftVar.first);
+    auto newFilterOutputVar = *objPool->add(new std::string("_NewFilter_"));
```
Comment (czpmango) [inline src/optimizer/rule/PushFilterDownLeftJoinRule.cpp:87]: genVar

File: src/validator/GoValidator.cpp [hunk 2]
```diff
@@ -759,7 +759,8 @@ Expression* GoValidator::rewriteToInputProp(const Expression* expr) {
     auto rewriter = [this](const Expression* e) -> Expression* {
         auto iter = propExprColMap_.find(e->toString());
         DCHECK(iter != propExprColMap_.end());
-        return new InputPropertyExpression(new std::string(*(iter->second->alias())));
+        return new VariablePropertyExpression(new std::string(""),
```
Comment (czpmango) [inline src/validator/GoValidator.cpp:762]: rename

File: src/optimizer/rule/PushFilterDownLeftJoinRule.cpp [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+            ->setCondition(objPool->add(filterUnpicked.release()));
+        oldLeftJoinNode->setInputVar(newFilterOutputVar);
+        oldLeftJoinNode->setLeftVar({newFilterOutputVar, 0});
+        const_cast<OptGroupNode*>(leftJoinGroupNode)->setDeps({newFilterGroup});
+        result.newGroupNodes.emplace_back(const_cast<OptGroupNode*>(filterGroupNode));
+    } else {
+        auto* newLeftJoinNode = static_cast<graph::LeftJoin*>(oldLeftJoinNode->clone());
+        newLeftJoinNode->setInputVar(newFilterOutputVar);
+        newLeftJoinNode->setOutputVar(oldFilterNode->outputVar());
+        newLeftJoinNode->setLeftVar({newFilterOutputVar, 0});
```
Comment (czpmango) [inline src/optimizer/rule/PushFilterDownLeftJoinRule.cpp:108]: Redundant data structures should be removed.

File: src/optimizer/rule/PushFilterDownLeftJoinRule.cpp [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+            DCHECK(expr->kind() == Expression::Kind::kVarProperty);
+            propNames.emplace_back(*static_cast<const VariablePropertyExpression*>(expr)->prop());
+        }
+        for (auto prop : propNames) {
+            if (std::find_if(
+                    leftVarColNames.begin(), leftVarColNames.end(), [&prop](std::string item) {
+                        return !item.compare(prop);
+                    }) == leftVarColNames.end()) {
+                return false;
+            }
```
Comment (yixinglu) [inline src/optimizer/rule/PushFilterDownLeftJoinRule.cpp:67]: it's more clear to use for-loop or define a variable for std::find_if

File: src/util/ExpressionUtils.cpp [hunk 5]
```diff
@@ -383,6 +400,60 @@ std::unique_ptr<Expression> ExpressionUtils::flattenInnerLogicalExpr(const Expre
...
+    const Expression *expr,
+    std::function<bool(const Expression *)> picker,
+    bool pickSwitch) {
+    // Pick the expressions that not contain LogicalAndExpr directly
+    if (expr->kind() != Expression::Kind::kLogicalAnd &&
+        !graph::ExpressionUtils::findAny(expr, {Expression::Kind::kLogicalAnd})) {
+        if (picker(expr) ^ !pickSwitch) {
+            return expr->clone();
+        } else {
+            return nullptr;
```
Comment (yixinglu) [inline src/util/ExpressionUtils.cpp:414]:
```
bool splitCondition(props, *pushedFilter, *notPushedFilter)
if (expr->kind() != LOGICLA) {
  if (check(expr)) {
    return expr;  
  }
  return nullptr;
}
e = flatten(expr);
for (op: e->operands) {
  if (check(op)) {
  } else {
  }
}

```

Comment (czpmango) [reply src/util/ExpressionUtils.cpp:414]: ACK.

File: src/optimizer/rule/PushFilterDownLeftJoinRule.cpp (hunk 4 above)
Comment (czpmango) [reply src/optimizer/rule/PushFilterDownLeftJoinRule.cpp:67]: ACK.

File: src/optimizer/rule/PushFilterDownLeftJoinRule.cpp [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+    }
+
+    auto newFilterOutputVar = newFilterNode->outputVar();
+    if (filterUnpicked) {
+        // oldFilter and oldLeftJoin need not be reconstructed here
+        static_cast<graph::Filter*>(oldFilterNode)
+            ->setCondition(objPool->add(filterUnpicked.release()));
+        oldLeftJoinNode->setLeftVar({newFilterOutputVar, 0});
+        const_cast<OptGroupNode*>(leftJoinGroupNode)->setDeps({newFilterGroup});
+        result.newGroupNodes.emplace_back(const_cast<OptGroupNode*>(filterGroupNode));
```
Comment (yixinglu) [inline src/optimizer/rule/PushFilterDownLeftJoinRule.cpp:103]: Generate new subplan rather than update this plan in-place.

Comment (czpmango) [reply src/optimizer/rule/PushFilterDownLeftJoinRule.cpp]: ACK.

**Review by yixinglu — APPROVED** [review]

**Review by jievince — APPROVED** [review]

**Pull request merged**